### PR TITLE
Bump to QEMU 5.2.0

### PR DESCRIPTION
**Pull request by wainersm**
This change the version of QEMU used in the tests and CI.

Not compiling with -O3 to avoid the warning:

Program python3 found: YES (/usr/bin/python3)
../meson.build:104: WARNING: Consider using the built-in optimization level instead of using "-O3".
../meson.build:108: WARNING: Consider using the built-in optimization level instead of using "-O3".

New and uneeded firmware files are removed, except for qboot.rom since it can be used with microvm
machine type.

Needs a patch which allows passing `--disable-tools --enable-virtiofsd`

Fixes: #1238

Signed-off-by: Wainer dos Santos Moschetta <wainersm@redhat.com>

### CODE REVIEW

**Comment by wainersm** [comment]
This is just a initial work, I still have to decide some questions as you can see on the issue linked. Let me run the CI to see what happens. :)

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (devimc) [inline tools/packaging/scripts/configure-hypervisor.sh:457]: please use the same "format":  `[[ condition1 || (condition2 && condition3)  ]]`

File: tools/packaging/static-build/qemu/build-static-qemu.sh
Comment (devimc) [inline tools/packaging/static-build/qemu/build-static-qemu.sh:60]: why this is needed ?

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (devimc) [inline tools/packaging/scripts/configure-hypervisor.sh:458]: why not just ignore the meson warning? what the default optimization?

**Review by devimc — COMMENTED** [review]
thanks @wainersm -  I've left some comments and questions

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (wainersm) [reply tools/packaging/scripts/configure-hypervisor.sh:458]: If we stick with `-O3` then this fix is require -> https://lists.gnu.org/archive/html/qemu-devel/2021-01/msg07262.html

By default QEMU is build with `-O2`. I'd like to understand if `-O3` is really required by Kata Containers.

File: tools/packaging/static-build/qemu/build-static-qemu.sh
Comment (wainersm) [reply tools/packaging/static-build/qemu/build-static-qemu.sh:60]: Ops, it's a leftover. I will remove it on final version of this PR.

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (devimc) [reply tools/packaging/scripts/configure-hypervisor.sh:458]: ok, this is is nice to have

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (jodh-intel) [inline tools/packaging/scripts/configure-hypervisor.sh:329]: Please use `bc` for these sorts of version comparision. We see bugs time and time again using this sort of logic. Here's an example (from this same script):

https://github.com/kata-containers/kata-containers/blob/main/tools/packaging/scripts/configure-hypervisor.sh#L253

**Review by jodh-intel — CHANGES_REQUESTED** [review]

**Comment by wainersm** [comment]
@devimc @jodh-intel guys, I just sent an updated version where I solely addressed your review. Thanks!

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (devimc) [inline tools/packaging/scripts/configure-hypervisor.sh:69]: `qemu_version` is local in `main`, how can you access it?, also afaik `bc` is unable to compare versions for exaple:

```sh
$ echo "5.0.0 > 2.7" | bc
(standard_in) 1: syntax error
```

Comment (jodh-intel) [reply tools/packaging/scripts/configure-hypervisor.sh:69]: Yep. I think you need to pass two version number strings and then either:

- Ensure both version numbers are in `major.minor` format and use `bc(1)`.
- Use @devimc's trick of using `sort -V` to handle semver version numbers if they are in `major.minor.patch`.

Comment (jodh-intel) [reply tools/packaging/scripts/configure-hypervisor.sh:69]: I wonder in fact whether it would make sense to write a `kata-version-compare.sh` script as it feels similar in some ways to https://github.com/kata-containers/tests/blob/main/.ci/kata-arch.sh (ideally with a set of bats unit tests ;)

However, I'm wondering if there is a rust/golang cli program that can do all this which we can standarise on (*surely* there is by now?!? :smile:)

Comment (jodh-intel) [reply tools/packaging/scripts/configure-hypervisor.sh:69]: https://github.com/timvisee/version-compare looks interesting, but doesn't provide a simple CLI which is odd :confused:.

Comment (dgibson) [reply tools/packaging/scripts/configure-hypervisor.sh:69]: Hmmm.  Parsing version numbers is nearly always a bad idea.  It very easily breaks if there are downstream versions of the package that have different features backported.

If at all possible, it's better to instead probe for the particular features that are required in the new version, rather than examining the version number itself.

Comment (wainersm) [reply tools/packaging/scripts/configure-hypervisor.sh:69]: @devimc , @jodh-intel , @dgibson , I appreciated your comments but I'd like to propose that the changes to `configure-hypervisor.sh` gets merged as is; then I open another issue for improving the script. is it reasonable?

Comment (devimc) [reply tools/packaging/scripts/configure-hypervisor.sh:69]: @wainersm - regardless @dgibson 's comment (good catch!), I think that your implementation of `qemu_version_is` is weak and buggy, isn't it? :thinking:

Comment (wainersm) [reply tools/packaging/scripts/configure-hypervisor.sh:69]: @devimc I just realize I was testing the script with a VERSION file were I put only "x.y". I fixed the script to correctly parse "x.y.z", ie., the `qemu_version` will be two digits "x.y". IMHO it is not needed to consider "z" on that script because it indicates the release of a stable branch (which in turn are security fixes only) which is unlikely to introduce new configure options.

**Comment by wainersm** [comment]
The new version of this PR fixes the problem with use of "x.y.z" with `bc` in the `configure-hypervisor.sh` script. Also it is added the patches which contain fixes for virtiofsd, as in https://github.com/kata-containers/kata-containers/pull/1360

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (devimc) [reply tools/packaging/scripts/configure-hypervisor.sh:69]: hmm... right... `calc` is still buggy but since you will work on the changes proposed by @dgibson 

lgtm

Comment (wainersm) [reply tools/packaging/scripts/configure-hypervisor.sh:69]: @devimc you mean buggy because passing a 3-digits version to the function then `bc` is gonna fail?

Comment (devimc) [reply tools/packaging/scripts/configure-hypervisor.sh:69]: yes

```
+ calc '2.0.0 >= 5'
+ local 'expr=2.0.0 >= 5'
++ echo 2.0.0 '>=' 5
++ bc
(standard_in) 1: syntax error
+ '[' '' == 1 ']'
```

Comment (wainersm) [reply tools/packaging/scripts/configure-hypervisor.sh:69]: Initially I was passing the `qemu_version` to `calc()` [1] as you mentioned above, @jodh-intel . But as I was repeating the same thing many times, I created the `qemu_version_is()`. I really don't have a strong preference here.

Also, before, the script was using local variables (`qemu_minor_version` and `qemu_major_version`) inside the method which were expected to be called from `main()`. So I really didn't change the semantic. I am also okay to promote `qemu_version`  global so easily accessible from anywhere in the script.

[1] btw, if you have a better name for this method then let me know

Comment (wainersm) [reply tools/packaging/scripts/configure-hypervisor.sh:69]: @jodh-intel @devimc Would the following change be enough?

```
diff --git a/tools/packaging/scripts/configure-hypervisor.sh b/tools/packaging/scripts/configure-hypervisor.sh
index ba87b7e7..b2350aac 100755
--- a/tools/packaging/scripts/configure-hypervisor.sh
+++ b/tools/packaging/scripts/configure-hypervisor.sh
@@ -38,6 +38,9 @@ typeset -A recognised_tags
 # Prefix were kata will be installed
 prefix=${PREFIX:-/usr}
 
+# The QEMU version on "major.minor" format.
+qemu_version=""
+
 recognised_tags=(
        [arch]="architecture-specific"
        [functionality]="required functionality"
@@ -55,13 +58,6 @@ calc() {
        [ "$(echo $expr | bc)" == "1" ]
 }
 
-# Display message to stderr and exit indicating script failed.
-die() {
-       local msg="$*"
-       echo >&2 "$script_name: ERROR: $msg"
-       exit 1
-}
-
 # Compare with the QEMU version.
 # Use a bc expression as $1.
 qemu_version_is() {
@@ -221,7 +217,7 @@ generate_qemu_options() {
 
        # Disabled options
 
-       if qemu_version_is ">= 5" ; then
+       if calc "${qemu_version} >= 5" ; then
                # Disable sheepdog block driver support
                qemu_options+=(size:--disable-sheepdog)
 
@@ -255,7 +251,7 @@ generate_qemu_options() {
 
        # Disable PAM authentication: it's a feature used together with VNC access
        # that's not used. See QEMU commit 8953caf for more details
-       qemu_version_is ">= 4" && qemu_options+=(size:--disable-auth-pam)
+       calc "${qemu_version} >= 4" && qemu_options+=(size:--disable-auth-pam)
 
        # Disable unused filesystem support
        [ "$arch" == x86_64 ] && qemu_options+=(size:--disable-fdt)
@@ -264,7 +260,7 @@ generate_qemu_options() {
        qemu_options+=(size:--disable-libnfs)
 
        # Starting from QEMU 4.1, libssh replaces to libssh2
-       if qemu_version_is ">= 4.1" ; then
+       if calc "${qemu_version} >= 4.1" ; then
                qemu_options+=(size:--disable-libssh)
        else
                qemu_options+=(size:--disable-libssh2)
@@ -297,7 +293,7 @@ generate_qemu_options() {
 
        # SECURITY: Don't build a static binary (lowers security)
        # needed if qemu version is less than 2.7
-       if qemu_version_is "< 2.7" ; then
+       if calc "${qemu_version} < 2.7" ; then
                qemu_options+=(security:--disable-static)
        fi
 
@@ -340,7 +336,7 @@ generate_qemu_options() {
        # But since QEMU 5.2 the daemon is built as part of the tools set
        # (disabled with --disable-tools) thus it needs to be explicitely
        # enabled.
-       if qemu_version_is ">= 5.2" ; then
+       if calc "${qemu_version} >= 5.2" ; then
                qemu_options+=(functionality:--enable-virtiofsd)
        fi
 
@@ -370,12 +366,12 @@ generate_qemu_options() {
        # Disable Capstone
        qemu_options+=(size:--disable-capstone)
 
-       if qemu_version_is ">= 3" ; then
+       if calc "${qemu_version} >= 3" ; then
                # Disable graphics
                qemu_options+=(size:--disable-virglrenderer)
 
                # Due to qemu commit 3ebb9c4f52, we can't disable replication in v3.0
-               if qemu_version_is "> 3" ; then
+               if calc "${qemu_version} > 3" ; then
                        # Disable block replication
                        qemu_options+=(size:--disable-replication)
                fi
@@ -399,7 +395,7 @@ generate_qemu_options() {
 
                # vxhs was deprecated on QEMU 5.1 so it doesn't need to be
                # explicitly disabled.
-               if qemu_version_is "< 5.1" ; then
+               if calc "${qemu_version} < 5.1" ; then
                        qemu_options+=(size:--disable-vxhs)
                fi
        fi
@@ -416,7 +412,7 @@ generate_qemu_options() {
 
        # Always strip binaries
        # needed if qemu version is less than 2.7
-       if qemu_version_is "< 2.7" ; then
+       if calc "${qemu_version} < 2.7" ; then
                qemu_options+=(size:--enable-strip)
        fi
 
@@ -433,12 +429,12 @@ generate_qemu_options() {
        qemu_options+=(functionality:--enable-cap-ng)
        qemu_options+=(functionality:--enable-seccomp)
 
-       if qemu_version_is ">= 3.1" ; then
+       if calc "${qemu_version} >= 3.1" ; then
                # AVX2 is enabled by default by x86_64, make sure it's enabled only
                # for that architecture
                if [ "$arch" == x86_64 ]; then
                        qemu_options+=(speed:--enable-avx2)
-                       if qemu_version_is ">= 5" ; then
+                       if calc "${qemu_version} >= 5" ; then
                                qemu_options+=(speed:--enable-avx512f)
                        fi
                        # According to QEMU's nvdimm documentation: When 'pmem' is 'on' and QEMU is
@@ -471,7 +467,7 @@ generate_qemu_options() {
        _qemu_cflags=""
 
        # compile with high level of optimisation
-       if qemu_version_is "< 5.2" ; then
+       if calc "${qemu_version} < 5.2" ; then
                _qemu_cflags+=" -O3"
        fi
```

Comment (devimc) [reply tools/packaging/scripts/configure-hypervisor.sh:69]: can you apply and push it? I think it needs more changes, for example `local qemu_version=` -> `qemu_version=` in `main`

Comment (wainersm) [reply tools/packaging/scripts/configure-hypervisor.sh]: @devimc good catch. also I deleted the `die()` method by mistake. Anyway, I just pushed another version of the commit.

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (devimc) [inline tools/packaging/scripts/configure-hypervisor.sh:63]: now you are using spaces to indent

**Review by devimc — COMMENTED** [review]
thanks @wainersm - one more nit :sweat_smile:

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (wainersm) [reply tools/packaging/scripts/configure-hypervisor.sh]: @devimc @jodh-intel , guys, I will drop the `bc` idea as it doesn't handle cases like '2.2 > 2.10' (return true, but should be false). I will send another implementation, where each digit is compared.

Comment (devimc) [reply tools/packaging/scripts/configure-hypervisor.sh]: good catch, try with `sort -V -C`

Comment (wainersm) [reply tools/packaging/scripts/configure-hypervisor.sh]: @devimc now I understood when @jodh-intel said you had a trick with `sort` :)

What do you think about this:
```
# Given $1 and $2 as version strings with 'x.y.z' format; if $1 >= $2 then
# return 0. Otherwise return 1.
# Use this function on conditionals to compare versions.
#
gt_eq() {
        format='^[0-9]+(\.[0-9]+)*$'
        if [[ ! ("$1" =~ $format && "$2" =~ $format) ]]; then
                echo "ERROR: Malformed version string"
        fi
        echo -e "$1\n$2" | sort -V -r -C
}
```

**Comment by skaegi** [comment]
We already have adopted QEMU 5.2.x in our own 1.x tree and this PR would be really helpful to us in terms of letting us move to 2.x. What is still left to do here? We're happy to test with it if this PR is mostly ready.

**Comment by wainersm** [comment]
> We already have adopted QEMU 5.2.x in our own 1.x tree and this PR would be really helpful to us in terms of letting us move to 2.x. What is still left to do here? We're happy to test with it if this PR is mostly ready.

Hi @skaegi , today I will send an updated version and I hope it will be the last for this PR. Then I will run the CI jobs. It would be nice if you could test on your use cases and report it here.

Another thing that should be worked out is the metrics CI since we don't want that this update introduce regressions on the metrics. For that I will have to catch up with @jcvenegas @devimc .

**Comment by Jakob-Naucke** [comment]
Hi @wainersm, can we add [this patch](https://lists.nongnu.org/archive/html/qemu-devel/2021-02/msg05793.html) please? It (re-)enables virtio-9p-ccw, required for 9p on s390x.

**Comment by wainersm** [comment]
Hi @Jakob-Naucke ,

> Hi @wainersm, can we add [this patch](https://lists.nongnu.org/archive/html/qemu-devel/2021-02/msg05793.html) please? It (re-)enables virtio-9p-ccw, required for 9p on s390x.

The patch applies clearly on 5.2 but the build fails because the `have_virtfs` variable is not defined (see below). I am passing explicitly `--enable-virtfs --enable-virtiofsd`. So the patch will need backport, or we cherry-pick the changes that introduced the virtfs configuration bits. Could you please work on that?

```
Program ../scripts/block-coroutine-wrapper.py found: YES

../hw/s390x/meson.build:43:3: ERROR: Unknown variable "have_virtfs".

A full log can be found at /root/qemu/build/meson-logs/meson-log.txt

ERROR: meson setup failed
```

**Comment by Jakob-Naucke** [comment]
@wainersm oh yes, sorry. Christian Ehrhardt of Canonical has already taken care of this for Ubuntu packaging and created a 5.2.0 compatible [backport](https://git.launchpad.net/ubuntu/+source/qemu/commit/?id=e4d879c593bc0f3758248de4e2b4087d9f5dfc7e)

**Comment by wainersm** [comment]
Hi @Jakob-Naucke ,

> @wainersm oh yes, sorry. Christian Ehrhardt of Canonical has already taken care of this for Ubuntu packaging and created a 5.2.0 compatible [backport](https://git.launchpad.net/ubuntu/+source/qemu/commit/?id=e4d879c593bc0f3758248de4e2b4087d9f5dfc7e)

I just updated this PR with that patch. I built QEMU locally but could not test (I rely on the kata-containers CI).

**Comment by wainersm** [comment]
@jcvenegas Hi Carlos, the jenkins-metrics-ubuntu-20-04 metrics job passed for this PR. I only would like to confirm that the results aren't not bogus so this PR can be merged once approved.

**Comment by wainersm** [comment]
Moved to WIP again because I need to fix the snap build...

**Comment by jcvenegas** [comment]
@wainersm I dont see any bogus result in  Metrics CI, @devimc @GabyCT  could confirm as well. The After fixing snap build this looks good to  me :)

**Comment by wainersm** [comment]
> @wainersm I dont see any bogus result in Metrics CI, @devimc @GabyCT could confirm as well. The After fixing snap build this looks good to me :)

Thanks for reviewing it @jcvenegas !

**Comment by wainersm** [comment]
Just updated this PR with a fix to snap build (commit 3202c03f5e7db79b2e73bcf037306f3a64e82c78)

**Comment by wainersm** [comment]
The jenkins-ci-ARM-ubuntu-18-04 build failed because it could not download muscl for arm. So nothing to so with this change.

**Review by fidencio — APPROVED** [review]
lgtm, thanks @wainersm!

I'd like to ask for @jcvenegas review here before having it merged.

**Comment by wainersm** [comment]
**One important note**:  This update will take effect for all architectures except for aarch64. According to versions.yaml, aarch64 has used QEMU 5.1 and I didn't change it. The reasons for not changing are:
 * I didn't know if the people interested on arm wanted that update
 * I didn't have hardware to test it and the arm CI job was broken (it seems fixed now)

I propose we merge this PR as is. If someone ask for aarch64 update then I can work on that but I will rely on the CI job for tests (so it should be working properly).

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (egernst) [inline tools/packaging/scripts/configure-hypervisor.sh:342]: To clarify; we normally have this disabled in the "configure-hyperivsor" script? Enabling it generally will add a loft of cruft we would otherwise not want?

**Comment by egernst** [comment]
Actively reviewing -- will try to leave some feedback by end of day.

I don't love having so many patches to carry. If at all feasible, we can push upstream to create a stable release?

**Review by egernst — APPROVED** [review]

**Comment by wainersm** [comment]
@egernst ,
 
> I don't love having so many patches to carry. If at all feasible, we can push the upstream folks to create a stable release?

Same here. And today I was inquiring the community on QEMU' irc channel about the process of stable releases. As you can see on the sniped below (note: only removed messages of people join/quit the channel) the process is somewhat ad-hoc. We can for sure ask for a new release with the CVE we care about.

Note that frequently people ask me "why not use the distro's QEMU?". I just sent an email to kata-dev because I don't know the rationale.

```
<wainersm> hi people! What's the cadence of the stable releases? Alongside the regular releases?
<danpb> wainersm: they're fairly arbitrary timeframe afaik
<wainersm> danpb, being up to the stable-branch maintainers to decide when?
<danpb> i assume they decide based on (a) when alot of stuff has accumulated  (b) when people request one is made for some important bug 
<pm215> as a practical observation, usually I think there is one .1 stable release which comes out broadly at the same time as the next .0 mainline release
<danpb> wainersm: is there a reason why you particularly care about the schedle ?
<wainersm> danpb, kata Containers uses upstream QEMU on its CI/tests (not sure, but some can use on production too). I'm working to update from 5.0 to 5.2 but I will need to carry CVE's patches. If we had a stable 5.2.1 release then I could just consume it. Since this situation can happen frequently, I'm interested on whether it had a regular cadence or not (for the sake of setting a process on kata side)
<pm215> note that "stable release" doesn't imply having patches for known CVE fixes
<danpb> wainersm: you should /not/ assume QEMU stable releases have CVE fixes
<danpb> QEMU essentially leaves this to downstream distros to solve
<pm215> if you want CVE fixes you need to either (a) use a distro QEMU or (b) monitor upstream QEMU/other CVE announcement locations and collate them yourself
<pm215> in particular, "we announced a CVE and there's a patch" does not trigger "we'll do a point release on the stable branch"
<danpb> wainersm: but that said, why does your CI environment need to have CVE fixes
<stsquad> wainersm: I'd use the distro release (for example Debian backports has a last stable release + fixes)
<stsquad> is my understanding correct that our (upstream) stable branches provide a starting point for the distro's final release?
<danpb> CI should be running in  self contained, throw away env generally, so what matters is security of the environment providing the CI env, not security of stuff under test
<wainersm> pm215, got it. and it makes sense
<wainersm> danpb, I don't have an answer for your question because the project had that approach before I joined. But I will ask the community and let you know :)
<danpb> wainersm: so you're saying that they currently run qemu 5.0 and backport CVE fixes for their CI env ?
<wainersm> stsquad, yep. I will bring that topic to the community. Don't know why it was chose to not use distro's QEMU.
<stsquad> the Linaro LKFT (kernel test) team have a similar requirement - they want a recent QEMU to run their tests in so they aren't pegged to very old versions in stable distros... but they don't want to track master (because having two moving targets is bad for CI)
<wainersm> danpb, yes, but just a few (4 or 5 patches) for virtiofsd
<wainersm> stsquad, yeah, track master wouldn't work for kata containers either (IMO)
<danpb> honestly for CI, i would always just track a distro
<danpb> there are bleeding edge distros if you need to be very close to latest, but using a stable distro is better for reliability IME
<wainersm> danpb, I will check if someone uses it on production-like environments 
<wainersm> hey danpb, stsquad, pm215, thanks for sharing those information!
<danpb> we really ought to mention this on the qemu.org download page
<stsquad> danpb: time for a flurry of parches for the website ;-)
<wainersm> danpb, https://qemu.readthedocs.io/en/latest/devel/stable-process.html has good information, maybe just put some complementary info? (IMHO it is not clear the cadence of the stable release- if any - that's why I asked)
<stsquad> the examples are a bit out of date on that so a refresh probably would help
```

**Comment by egernst** [comment]
Thanks @wainersm  -- let's wait for more feedback on ML, then post summary here for reference (I agree with Julio's feedback).

**Review by devimc — APPROVED** [review]

**Pull request merged**
